### PR TITLE
Add support for ms, mr and ma annotations

### LINKED ISSUE #150: Add support for Message Broker type messages
Open Zipkin documentation defines ms/mr type of messages for cases with a publisher and one or more consumers of a given event. Unfortunately zipkin4net currently does not support sending these types of spans to the server. 

I have tested submitting a JSON request via v1/api/spans and Zipkin UI renders associated events. 

    [
        {
            "traceId": "f95ae2017821cb14",
            "id": "886a8994ecf8bf35",
            "name": "consumer1method",
            "timestamp": 1507247375925111,
            "annotations": [
                {
                    "timestamp": 1507247375925111,
                    "value": "mr",
                    "endpoint": {
                        "serviceName": "consumer1"
                    }
                }
            ],
            "binaryAnnotations": [
                {
                    "key": "ID",
                    "value": "76a5ef53-61c6-4043-8ee5-1f74fa2dd6b1",
                    "endpoint": {
                        "serviceName": "consumer1"
                    }
                }
            ]
        },
        {
            "traceId": "f95ae2017821cb14",
            "id": "87fd0e42c0cc0129",
            "name": "producerprocess",
            "parentId": "886a8994ecf8bf35",
            "timestamp": 1507247376027131,
            "annotations": [
                {
                    "timestamp": 1507247376027131,
                    "value": "ms",
                    "endpoint": {
                        "serviceName": "producer"
                    }
                }
            ]
        },
        {
            "traceId": "f95ae2017821cb14",
            "id": "8094558e4feeaf12",
            "name": "consumer2process",
            "parentId": "87fd0e42c0cc0129",
            "timestamp": 1507247376047135,
            "annotations": [
                {
                    "timestamp": 1507247376047135,
                    "value": "mr",
                    "endpoint": {
                        "serviceName": "consumer2"
                    }
                }
            ]
        }
    ]

It would be very helpful if this was added to the interface to support these types of events as there is a lot of distributed systems now that rely on pub/sub type of flows. Thank you.

### PR DESCRIPTION
**Pull request by fedj**
- Update thrift files
- Add ms, mr and ma support

Fixes #150

### CODE REVIEW

**Comment by fedj** [comment]
cc @crntn

**Comment by aachinfiev** [comment]
You guys are awesome. Fast response and implementation :) Question about the code change: you support message send start/stop which suppose to record duration, however when publisher sends the request they just issue a single publish. There is no response for which you would call a stop on.

From zipkin docs I thought you would just have two events: MessageSend, MessageReceive. Could you please explain what you are using the stops for to compute duration? Thank you.

**Comment by fedj** [comment]
@aachinfiev Indeed it can look confusing but the duration is not really computed.
If you take a look [here](https://github.com/openzipkin/zipkin4net/blob/master/Src/zipkin4net/Src/Tracers/Zipkin/Span.cs#L92), you'll see that this is not a supported case for duration computation.

MessageSend and MessageRecv are splitted in Start and Stop due to the current design. It allows to add binary annotations such as "ma" or add wire send for example.
I think that v2 model will ease this and will remove the need of such tricks.

Basically, you'll need to call MessageSendStart, add MessageAddr, publish it and then call MessageSendStop.

Is it clearer ?

**Comment by aachinfiev** [comment]
Oh I see. Yes that makes sense. Thanks.

**Comment by codefromthecrypt** [comment]
Looks like a good start.. almost wonder if we should do this in v1 model at
all :)

Since the annotations arent exact match anyway, maybe ProducerStart/End?
same for consumer?

Then you can use the mapping to v1 underneath. Note there are very specific
mappings of this to wire annotation.. wire annotation is only to allow
duration which is optional on messaging spans.

https://github.com/openzipkin/zipkin/blob/master/zipkin/src/main/java/zipkin/internal/V2SpanConverter.java#L382

https://github.com/openzipkin/zipkin/blob/master/zipkin/src/test/java/zipkin/internal/V2SpanConverterTest.java#L493

**Comment by fedj** [comment]
@adriancole Good point, done.

File: Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs [hunk 1]
```diff
@@ -145,15 +146,28 @@ public void Read (TProtocol iprot)
...
-      sb.Append(",Value: ");
-      sb.Append(Value);
-      sb.Append(",Host: ");
-      sb.Append(Host== null ? "<null>" : Host.ToString());
-      sb.Append(")");
-      return sb.ToString();
+      StringBuilder __sb = new StringBuilder("Annotation(");
+      bool __first = true;
+      if (Timestamp != null) {
+        if(!__first) { __sb.Append(", "); }
```
Comment (crntn) [inline Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs]: Can remove, always true

File: Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs [hunk 2]
```diff
@@ -145,15 +146,28 @@ public void Read (TProtocol iprot)
...
+      }
+      if (Value != null) {
+        if(!__first) { __sb.Append(", "); }
+        __first = false;
+        __sb.Append("Value: ");
+        __sb.Append(Value);
+      }
+      if (Host != null) {
+        if(!__first) { __sb.Append(", "); }
+        __first = false;
```
Comment (crntn) [inline Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs]: can remove too, this is never used

File: Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs [hunk 3]
```diff
@@ -145,15 +146,28 @@ public void Read (TProtocol iprot)
...
-      sb.Append("Timestamp: ");
-      sb.Append(Timestamp);
-      sb.Append(",Value: ");
-      sb.Append(Value);
-      sb.Append(",Host: ");
-      sb.Append(Host== null ? "<null>" : Host.ToString());
-      sb.Append(")");
-      return sb.ToString();
+      StringBuilder __sb = new StringBuilder("Annotation(");
+      bool __first = true;
```
Comment (crntn) [inline Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs]: local variable should not start with _

File: Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs (hunk 1 above)
Comment (crntn) [inline Src/zipkin4net/Src/Tracers/Zipkin/thrift_gen/gen-csharp/zipkin4net/Tracers/Zipkin/Thrift/Annotation.cs:152]: weird code generated by thrift

**Review by crntn — APPROVED** [review]
